src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs

```diff
@@ -55,36 +50,61 @@ internal class AnalysisScope
         public bool IsSingleFileAnalysis => FilterFileOpt != null;
 
         /// <summary>
-        /// Flag indicating if this is a partial analysis for the corresponding <see cref="CompilationWithAnalyzers"/>,
-        /// i.e. <see cref="IsSingleFileAnalysis"/> is true and/or <see cref="Analyzers"/> is a subset of <see cref="CompilationWithAnalyzers.Analyzers"/>.
+        /// Flag indicating if this analysis scope contains all analyzers from the corresponding <see cref="CompilationWithAnalyzers"/>,
+        /// i.e. <see cref="Analyzers"/> is the same set as <see cref="CompilationWithAnalyzers.Analyzers"/>.
+        /// This flag is used to improve the performance for <see cref="Contains(DiagnosticAnalyzer)"/> check for
+        /// batch compilation scenario, where this flag is always true.
+        /// </summary>
+        private bool HasAllAnalyzers { get; }
+
+        /// <summary>
+        /// True if we are performing syntactic or semantic analysis for a single source file with a single analyzer in scope,
+        /// which is a <see cref="CompilerDiagnosticAnalyzer"/>.
         /// </summary>
-        public bool IsPartialAnalysis { get; }
+        public bool IsSingleFileAnalysisForCompilerAnalyzer =>
+            IsSingleFileAnalysis && Analyzers is [CompilerDiagnosticAnalyzer];
 
         /// <summary>
         /// True if we are performing semantic analysis for a single source file with a single analyzer in scope,
         /// which is a <see cref="CompilerDiagnosticAnalyzer"/>.
         /// </summary>
         public bool IsSemanticSingleFileAnalysisForCompilerAnalyzer =>
-            IsSingleFileAnalysis && !IsSyntacticSingleFileAnalysis && Analyzers is [CompilerDiagnosticAnalyzer];
+            IsSingleFileAnalysisForCompilerAnalyzer && !IsSyntacticSingleFileAnalysis;
 
-        public AnalysisScope(Compilation compilation, AnalyzerOptions? analyzerOptions, ImmutableArray<DiagnosticAnalyzer> analyzers, bool hasAllAnalyzers, bool concurrentAnalysis, bool categorizeDiagnostics)
-            : this(compilation.SyntaxTrees, analyzerOptions?.AdditionalFiles ?? ImmutableArray<AdditionalText>.Empty,
-                   analyzers, isPartialAnalysis: !hasAllAnalyzers, filterFile: null, filterSpanOpt: null, isSyntacticSingleFileAnalysis: false, concurrentAnalysis: concurrentAnalysis, categorizeDiagnostics: categorizeDiagnostics)
+        public static AnalysisScope Create(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, CompilationWithAnalyzers compilationWithAnalyzers)
```

Comment:
Replaced the public ctors with static `Create` methods to reduce the data that callers need to pass in.